webhooks/clusterclass.go

```diff
@@ -123,8 +140,126 @@ func (webhook *ClusterClass) validate(old, new *clusterv1.ClusterClass) error {
 	// Ensure spec changes are compatible.
 	allErrs = append(allErrs, check.ClusterClassesAreCompatible(old, new)...)
 
+	// Ensure no MachineDeploymentClass currently in use has been removed from the ClusterClass.
+	allErrs = append(allErrs, webhook.validateLivingMachineDeploymentClassesNotRemoved(old, new)...)
+
 	if len(allErrs) > 0 {
 		return apierrors.NewInvalid(clusterv1.GroupVersion.WithKind("ClusterClass").GroupKind(), new.Name, allErrs)
 	}
 	return nil
 }
+
+func (webhook *ClusterClass) validateLivingMachineDeploymentClassesNotRemoved(old, new *clusterv1.ClusterClass) field.ErrorList {
+	var allErrs field.ErrorList
+	if old == nil {
+		return allErrs
+	}
+	missingClasses := webhook.removedMachineClasses(old, new)
+
+	// If no classes have been removed return early as no further checks are needed.
+	if len(missingClasses) == 0 {
+		return allErrs
+	}
+
+	// TODO: The first part of this method will need to be broken out to properly account for changes when also validating variables.
+	// Retrieve all clusters using the ClusterClass.
+	clusters, errs := webhook.clustersUsingClusterClass(old)
+	if len(errs) > 0 {
+		allErrs = append(allErrs, errs...)
+		return allErrs
+	}
+
+	// Retrieve all the MachineDeployments in the clusterClass namespace using a managed topology.
+	machineDeployments, errs := webhook.topologyManagedMachineDeployments(new.Namespace)
```

Comment:
I'm not sure if we have to access the MDs. I think checking the MD topologies in the Cluster is enough.

Reply:
There's a tradeoff here on which race conditions we want to account for I think. If we just check the cluster topologies there may be some impact on existing machineDeployments on being reconciled before deletion.

Just checking the cluster object makes the code a lot simpler and more efficient though.

Reply:
let's start with the simplest solution and rely on the Cluster only

Reply:
@fabriziopandini We verified it together (by setting up a tilt cluster + quickstart e2e + manual modifications), we're safe against race conditions and don't have to look at MDs in this webhook.

Concrete:
* Starting point: ClusterClass and Cluster have a MDClass/MDTopology
* After dropping the MDClass from the ClusterClass the cluster reconcile fails (as expected)
* After dropping the MDTopology from the Cluster the cluster reconcile is able to delete the MD (it doesn't have to access the missing MDClass to do that)